Makefile

```
@@ -4,10 +4,18 @@ IMAGE_NAME=vault-csi-provider
 VERSION?=1.0.0
 IMAGE_TAG=$(REGISTRY_NAME)/$(IMAGE_NAME):$(VERSION)
 IMAGE_TAG_LATEST=$(REGISTRY_NAME)/$(IMAGE_NAME):latest
-BUILD_DATE=$$(date +%Y-%m-%d-%H:%M)
-LDFLAGS?="-X 'github.com/hashicorp/vault-csi-provider/internal/version.BuildVersion=$(VERSION)' \
-	-X 'github.com/hashicorp/vault-csi-provider/internal/version.BuildDate=$(BUILD_DATE)' \
-	-X 'github.com/hashicorp/vault-csi-provider/internal/version.GoVersion=$(shell go version)'"
+# https://reproducible-builds.org/docs/source-date-epoch/
+DATE_FMT= +'%Y-%m-%d-%H:%M'
+SOURCE_DATE_EPOCH ?= $(shell git log -1 --pretty=%ct)
+ifdef SOURCE_DATE_EPOCH
+    BUILD_DATE ?= $(shell date -u -d "@$(SOURCE_DATE_EPOCH)" "$(DATE_FMT)" 2>/dev/null || date -u -r "$(SOURCE_DATE_EPOCH)" "$(DATE_FMT)" 2>/dev/null || date -u "$(DATE_FMT)")
```

Running `make build` gave me some quoting issues, as the BUILD_DATE `-X` option ends up like this:

```text
-X 'github.com/hashicorp/vault-csi-provider/internal/version.BuildDate='2022-02-03-21:38''
```

On an ubuntu container:

```shell
$ date -u -d "@1643924308" "+'%Y-%m-%d-%H:%M'"
'2022-02-03-21:38'
$ date -u -d "@1643924308" +'%Y-%m-%d-%H:%M'
2022-02-03-21:38
```

On Mac:

```shell
$ date -u -r "1643924308" "+'%Y-%m-%d-%H:%M'"
'2022-02-03-21:38'
$ date -u -r "1643924308" +'%Y-%m-%d-%H:%M'
2022-02-03-21:38
```

I'm no expert on quoting, but does the following suggestion seem reasonable? It fixes `make build` for me.

```suggestion
    BUILD_DATE ?= $(shell date -u -d "@$(SOURCE_DATE_EPOCH)" $(DATE_FMT) 2>/dev/null || date -u -r "$(SOURCE_DATE_EPOCH)" $(DATE_FMT) 2>/dev/null || date -u "$(DATE_FMT)")
```

---

I also updated code according to your review, thanks a lot, yes we should remove double quotes around DATE_FMT also.

---

I did the same for aws/karpenter repository, please see (I wrapped DATE_FMT with quotes, nothing wrong happened)
https://github.com/aws/karpenter/commit/dc72a87bc71e2f3169392edd44c132d2efa73ca7